openwisp_monitoring/device/apps.py

```diff
@@ -11,3 +15,57 @@ class DeviceMonitoringConfig(AppConfig):
 
     def ready(self):
         manage_short_retention_policy()
+        from openwisp_controller.connection.models import DeviceConnection
+        from openwisp_controller.connection.signals import is_working_changed
+        is_working_changed.connect(self.health_status_changed,
+                                   sender=DeviceConnection,
+                                   dispatch_uid='change_health_status')
+
+    @classmethod
+    def health_status_changed(cls, instance, is_working, **kwargs):
+        from ..check.models import Check
+        d = instance.device
+        dm = d.monitoring
+        initial_status = dm.status
+        status = 'ok' if is_working else 'problem'
+        if not is_working:
+            dm.update_status(status)
+            # Create a related notification explaining why it's not working
+            desc = instance.failure_reason
+            opts = dict(actor=d, level='warning', verb='not working', description=desc)
+            if initial_status == 'ok':
+                try:
+                    # if checks exist trigger them
+                    Check.objects.get(object_id=d.pk)
+                    management.call_command('run_checks')
+                except Check.DoesNotExist:
+                    pass
+        else:
+            # create a notification that device is working
+            opts = dict(actor=d, level='info', verb='connected successfully')
+            try:
+                # if checks exist trigger them else, set status as 'ok'
+                Check.objects.get(object_id=instance.device.pk)
```

Comment:
This will run checks on all devices instead of only this one. Please correct me if I am missing something.

Reply:
Yep, it shouldn't be the case but I have kept it for now. The reason being, I have created a new task for #14 and will use the same here later on. I wanted to avoid duplication so did away with them here.